extensions/csrc/cuda/attention/attention_utils.h

```diff
@@ -0,0 +1,504 @@
+/*
+ * Adapted from
+ * https://github.com/NVIDIA/FasterTransformer/blob/release/v5.3_tag/src/fastertransformer/kernels/decoder_masked_multihead_attention/decoder_masked_multihead_attention_template.hpp
+ * Copyright (c) 2024, The Colossal-AI team.
+ * Copyright (c) 2023, The vLLM team.
+ * Copyright (c) 2020-2023, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cuda_bf16.h>
+#include <cuda_fp16.h>
+#include <float.h>
+
+#include "attention_generic.h"
+#include "cuda_type_utils.h"
+
+// Define custom BF16 vector data types.
+struct bfloat164 {
+  __nv_bfloat162 x;
+  __nv_bfloat162 y;
+};
+
+struct bfloat168 {
+  __nv_bfloat162 x;
+  __nv_bfloat162 y;
+  __nv_bfloat162 z;
+  __nv_bfloat162 w;
+};
+
+struct half4 {
+  half2 x;
+  half2 y;
+};
+
+struct half8 {
+  half2 x;
+  half2 y;
+  half2 z;
+  half2 w;
+};
+
+template <>
+struct VecType<__nv_bfloat16, 2> {
+  using Type = __nv_bfloat162;
+};
+template <>
+struct VecType<__nv_bfloat16, 4> {
+  using Type = bfloat164;
+};
+template <>
+struct VecType<__nv_bfloat16, 8> {
+  using Type = bfloat168;
+};
+
+template <>
+struct VecType<half, 2> {
+  using Type = half2;
+};
+template <>
+struct VecType<half, 4> {
+  using Type = half4;
+};
+template <>
+struct VecType<half, 8> {
+  using Type = half8;
+};
+
+template <>
+struct VecType<float, 2> {
+  using Type = float2;
+};
+template <>
+struct VecType<float, 4> {
+  using Type = float4;
+};
+template <>
+struct VecType<float, 8> {
+  using Type = float8_;
```

Comment:
Although these are expected to be only used by attention, there exist some duplicate ones with type traits in `extensions/csrc/cuda/utils/vec_type_traits.h`. You might want to compare the two files and clear up a little.